8252739: Deflater.setDictionary(byte[], int off, int len) ignores the starting offset for the dictionary

LanceAndersen opened the pull request:
Hi all,

Please review the fix  for JDK-8252739 which addresses an issue introduced by https://bugs.openjdk.java.net/browse/JDK-8225189, where Deflater.c ignored the offset specified by Deflater.setDictionary.

Mach5 jdk-tier1, jdk-tier2, jdk-tier3 runs cleanly as well as the java/util/zip and java/util/jar JCK tests.
<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252739](https://bugs.openjdk.java.net/browse/JDK-8252739): Deflater.setDictionary(byte[], int off, int len) ignores the starting offset for the dictionary


### Reviewers
 * [Uwe Schindler](https://openjdk.java.net/census#uschindler) (@uschindler - Author) ⚠️ Review applies to aac03e3e8d4ef4a7ca1a8e715b4ec62a68e08d35
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/269/head:pull/269`
`$ git checkout pull/269`

uschindler commented on test/jdk/java/util/zip/DeflaterDictionaryTests.java:
```diff
@@ -0,0 +1,185 @@
+        byte[] input = SRC_DATA.getBytes(UTF_8);
+        byte[] output = new byte[RESULT_SIZE];
+        // Compress the bytes
+        ByteBuffer dictDef = ByteBuffer.wrap(DICTIONARY.getBytes(UTF_8));
```
This should use a slice of the byte buffer, so array offset is not 0. This can be done with offset and size on the wrap() call.

uschindler commented on test/jdk/java/util/zip/DeflaterDictionaryTests.java:
```diff
@@ -0,0 +1,185 @@
+        byte[] input = SRC_DATA.getBytes(UTF_8);
+        byte[] output = new byte[RESULT_SIZE];
+        // Compress the bytes
+        ByteBuffer dictDef = ByteBuffer.allocateDirect(DICTIONARY.length());
```
This creates an empty dictionary with only 0 bytes. Maybe copy some real data into the buffer. Also a slice should be used.

uschindler reviewed (changes requested):
The tests with byte buffers (direct and heap) are not using offsets (arrayOffset=0). The direct buffer test uses just a series of 0-bytes, so incorrect offsets won't change result. There should be real data copied into the direct buffer.

LanceAndersen commented:
Minor updates have been made to the tests

uschindler commented:
Ok much better for the heap buffers. Many thanks.

The direct buffers have now contents, but I think it should copy the whole heap array into the byte buffer. After that set position and limit and then call slice(). After that you have a slice of the original direct buffer with offset.

LanceAndersen commented:
I have made additional updates to testByteBufferDirect to incorporate your suggestions

uschindler reviewed (approved):
Looks fine to me.

AlanBateman reviewed (commented):
Update to Deflater.c looks good.

DeflaterDictionaryTests looks like is a shaping up to be a good test for setDictionary. Are there other assertions that should be checked, e.g. setDictionary(ByteBuffer) is specified to consume all bytes and it would be good to check that the position is set to the limit. Also can the 2-arg setDictionary be tests, also corner cases such no bytes remaining or invoked with null.

The naming of the tests is a bit inconsistent, e.g. ByteArrayTest vs. testByteBufferDirect. In the naming then I would use "Heap" instead of "Wrap", as in testHeapByteBuffer.

In passing: The tests can use try-finally to ensure that Deflater::end is invoked even when the test fails. String repeat(int) could be used to avoid duplicating "Welcome to Us Open;". Missing space in several "if(...)" usages.

LanceAndersen commented:
> Update to Deflater.c looks good.
> 
> DeflaterDictionaryTests looks like is a shaping up to be a good test for setDictionary. Are there other assertions that should be checked, e.g. setDictionary(ByteBuffer) is specified to consume all bytes and it would be good to check that the position is set to the limit. Also can the 2-arg setDictionary be tests, also corner cases such no bytes remaining or invoked with null.
> 

I have logged JDK-8253444 to add more coverage for setDictionary.  I would prefer to put his change back to address the offset not being used and improve the coverage afterwards.

> The naming of the tests is a bit inconsistent, e.g. ByteArrayTest vs. testByteBufferDirect. In the naming then I would use "Heap" instead of "Wrap", as in testHeapByteBuffer.

I  updated the test name(s).
> 
> In passing: The tests can use try-finally to ensure that Deflater::end is invoked even when the test fails. String repeat(int) could be used to avoid duplicating "Welcome to Us Open;". Missing space in several "if(...)" usages.

Addressed the issues above.

I have pushed the updated changes for review to the PR.

AlanBateman reviewed (commented):
Given that the offset handling is buggy then I think the test should at least cover the cases where the offset is 0 or out of bounds. No problem separating out further test coverage for the other setDictionary methods into a separate issue.

LanceAndersen commented:
> Given that the offset handling is buggy then I think the test should at least cover the cases where the offset is 0 or out of bounds. No problem separating out further test coverage for the other setDictionary methods into a separate issue.

I updated the test to include an offset test for 0 and to add a test for out of range offsets.  I have updated the PR.

AlanBateman reviewed (approved):
Thanks for update to cover the offset 0 and out of bound cases.

LanceAndersen commented:
/integrate
